fvm/programs/programs.go

```diff
@@ -27,73 +27,57 @@ type ProgramEntry struct {
 	State    *state.State
 }
 
-type ProgramGetFunc func(location common.Location) (*ProgramEntry, bool)
-
-func emptyProgramGetFunc(_ common.Location) (*ProgramEntry, bool) {
-	return nil, false
-}
-
 // Programs is a cumulative cache-like storage for Programs helping speed up execution of Cadence
 // Programs don't evict elements at will, like a typical cache would, but it does it only
 // during a cleanup method, which must be called only when the Cadence execution has finished.
 // It it also fork-aware, support cheap creation of children capturing local changes.
 type Programs struct {
-	lock       sync.RWMutex
-	programs   map[common.LocationID]ProgramEntry
-	parentFunc ProgramGetFunc
-	cleaned    bool
+	lock     sync.RWMutex
+	programs map[common.LocationID]*ProgramEntry
+	parent   *Programs
+	cleaned  bool
 }
 
 func NewEmptyPrograms() *Programs {
-
 	return &Programs{
-		programs:   map[common.LocationID]ProgramEntry{},
-		parentFunc: emptyProgramGetFunc,
+		programs: map[common.LocationID]*ProgramEntry{},
 	}
 }
 
 func (p *Programs) ChildPrograms() *Programs {
 	return &Programs{
-		programs: map[common.LocationID]ProgramEntry{},
-		parentFunc: func(location common.Location) (*ProgramEntry, bool) {
-			return p.get(location)
-		},
+		programs: map[common.LocationID]*ProgramEntry{},
+		parent:   p,
 	}
 }
 
 // Get returns stored program, state which contains changes which correspond to loading this program,
 // and boolean indicating if the value was found
 func (p *Programs) Get(location common.Location) (*interpreter.Program, *state.State, bool) {
-	p.lock.RLock()
-	defer p.lock.RUnlock()
-
-	programEntry, has := p.get(location)
+	entry, parent := p.get(location)
+	if entry != nil {
+		return entry.Program, entry.State, true
+	}
 
-	if has {
-		return programEntry.Program, programEntry.State, true
+	if parent != nil {
+		return parent.Get(location)
 	}
 
 	return nil, nil, false
 }
 
-func (p *Programs) get(location common.Location) (*ProgramEntry, bool) {
-	programEntry, ok := p.programs[location.ID()]
-	if !ok {
-		parentEntry, has := p.parentFunc(location)
+func (p *Programs) get(location common.Location) (*ProgramEntry, *Programs) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
 
-		if has {
-			return parentEntry, true
-		}
-		return nil, false
-	}
-	return &programEntry, true
+	return p.programs[location.ID()], p.parent
 }
 
 func (p *Programs) Set(location common.Location, program *interpreter.Program, state *state.State) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	p.programs[location.ID()] = ProgramEntry{
+	p.programs[location.ID()] = &ProgramEntry{
```

Comment:
Was there a reason to move to a pointer? Isn't this increasing allocations count, instead of keeping data inside a map?